runtime/near-vm-logic/src/logic.rs

```diff
@@ -109,88 +133,162 @@ impl<'a> VMLogic<'a> {
+        let data = self.memory_get_vec(offset, memory_len)?;
+        let mut res = vec![0u64; num_elements as usize];
+        byteorder::LittleEndian::read_u64_into(&data, &mut res);
+        Ok(res)
     }
 
-    fn read_memory_u32(memory: &dyn MemoryLike, ptr: u64) -> Result<u32> {
-        let mut slice = [0u8; size_of::<u32>()];
-        let buf = Self::memory_get(memory, ptr, size_of::<u32>() as u64)?;
-        slice.copy_from_slice(&buf);
-        Ok(u32::from_le_bytes(slice))
+    fn get_vec_from_memory_or_register(&mut self, offset: u64, len: u64) -> Result<Vec<u8>> {
+        if len != std::u64::MAX {
+            self.memory_get_vec(offset, len)
+        } else {
+            self.internal_read_register(offset)
+        }
     }
 
-    fn get_from_memory_or_register(
-        memory: &dyn MemoryLike,
-        registers: &HashMap<u64, Vec<u8>>,
-        offset: u64,
-        len: u64,
-    ) -> Result<Vec<u8>> {
-        if len != std::u64::MAX {
-            Self::memory_get(memory, offset, len)
+    fn memory_set_slice(&mut self, offset: u64, buf: &[u8]) -> Result<()> {
+        self.gas_counter.pay_base(write_memory_base)?;
+        self.gas_counter.pay_per_byte(write_memory_byte, buf.len() as _)?;
+        self.try_fit_mem(offset, buf.len() as _)?;
+        self.memory.write_memory(offset, buf);
+        Ok(())
+    }
+
+    memory_set!(u128, memory_set_u128);
+
+    // #################
+    // # Registers API #
+    // #################
+
+    fn internal_read_register(&mut self, register_id: u64) -> Result<Vec<u8>> {
+        if let Some(data) = self.registers.get(&register_id) {
+            self.gas_counter.pay_base(read_register_base)?;
+            self.gas_counter.pay_per_byte(read_register_byte, data.len() as _)?;
+            Ok(data.clone())
+        } else {
+            Err(HostError::InvalidRegisterId.into())
+        }
+    }
+
+    fn internal_write_register(&mut self, register_id: u64, data: Vec<u8>) -> Result<()> {
+        self.gas_counter.pay_base(write_register_base)?;
+        self.gas_counter.pay_per_byte(write_register_byte, data.len() as u64)?;
+        if data.len() as u64 > self.config.max_register_size
+            || self.registers.len() as u64 >= self.config.max_number_registers
+        {
+            return Err(HostError::MemoryAccessViolation.into());
+        }
+        self.registers.insert(register_id, data);
+
+        // Calculate the new memory usage.
+        let usage: usize =
+            self.registers.values().map(|v| size_of::<u64>() + v.len() * size_of::<u8>()).sum();
+        if usage as u64 > self.config.registers_memory_limit {
+            Err(HostError::MemoryAccessViolation.into())
         } else {
-            registers.get(&offset).ok_or(HostError::InvalidRegisterId.into()).map(|v| v.clone())
+            Ok(())
         }
     }
 
+    /// Convenience function for testing.
+    pub fn wrapped_internal_write_register(&mut self, register_id: u64, data: &[u8]) -> Result<()> {
```

Comment:
Better to hide it behind a #[cfg(test)]

Reply:
This will not work with integration tests since they compile `near-vm-logic` without `test` feature.